optimade/models/utils.py

```diff
@@ -1,8 +1,108 @@
+import inspect
+import warnings
 import re
+from enum import Enum
+from typing import Optional
+
+from pydantic import Field
+
+_PYDANTIC_FIELD_KWARGS = list(inspect.signature(Field).parameters.keys())
 
 __all__ = ("CHEMICAL_SYMBOLS", "EXTRA_SYMBOLS", "ATOMIC_NUMBERS", "SemanticVersion")
 
 
+class SupportLevel(Enum):
+    MUST = "must"
+    SHOULD = "should"
+    OPTIONAL = "optional"
+
+
+def StrictField(
+    *args,
+    description: str = None,
+    **kwargs,
+) -> Field:
+    """A wrapper around `pydantic.Field` that does the following:
+
+    - Forbids any "extra" keys that would be passed to `pydantic.Field`,
+      except those used elsewhere to modify the schema in-place,
+      e.g. "uniqueItems", "pattern" and those added by OptimadeField, e.g.
+      "unit", "queryable" and "sortable".
+    - Emits a warning when no description is provided.
+
+    Raises:
+        RuntimeError: if a Field is created with an unexpected key.
+
+    Returns:
+        the pydantic field
+
+    """
+
+    allowed_keys = [
+        "unit",
+        "pattern",
+        "uniqueItems",
+        "support",
+        "queryable",
+        "sortable",
+    ]
+    _banned = [k for k in kwargs if k not in set(_PYDANTIC_FIELD_KWARGS + allowed_keys)]
+
+    if _banned:
+        raise RuntimeError(
+            f"Not creating StrictField({args}, {kwargs}) with forbidden keywords {_banned}."
+        )
+
+    if description is not None:
+        kwargs["description"] = description
+
+    if description is None:
```

Comment:
Is not working for description? That's weird. Maybe remove it in the input arguments and add it to allowed keys?

Reply:
Whats the problem here? I wanted to add `description` on a different level to the other keys so that it appears in the docstring for `StrictField`

Reply:
Right, okay. So it's essentially not just an "allowed" key, but a mandatory key?

Reply:
At the risk of mixing terminology, I would say its a SHOULD key :P The main difference being that you get a warning if you miss it out from the `StrictField`. I guess you could argue that it should be mandatory for an `OptimadeField`?

Reply:
Hehe. Fair enough. But why not just make it part of the `kwargs` then? Keeping the documentation in the doc string (under the `Keyword arguments` headline) and then adding another "should" list to go through. Or just leave it as is, warning if `description` is not found in `kwargs`?

Reply:
The reason why I'm pushing a bit, I think, is because it seems very hard-coded all the `description`-specific code lines here. And it should perhaps be less so. But if it's the only "should" level argument at the moment.. well...

Reply:
The reason it's hardcoded is because its the main point of this wrapper; I think its better to be explicit in the function signature that it is in some sense "expected". Re: putting it in the docstring, it looks like mkdocstrings doesn't nicely handle keyword arguments blocks, so it'd be a messy doctstring about **kwargs having to contain description etc... I could add what I'm saying here to the docstring itself to make it clearer I guess